### ISSUE TITLE
bulkio: investigate “long tail” index backfill performance

### ISSUE THREAD

**Issue by shermanCRL** [cockroachdb/cockroach#67535]
When working with a client, we observed a stalled, or at least very slow, index creation. Part of the problem was slowness of the Jobs table, addressed in a separate issue [here](https://github.com/cockroachdb/cockroach/issues/67523).

After unblocking that issue, index backfill succeeded but appeared to have a “long tail”, in that the last “two minutes” took several hours. We also observed that the hardware was underutilized -- general low CPU and not as much distribution across nodes as we might hope for.

Let’s investigate:

- Is there a “long tail” performance issue, i.e., does it slow down over time? Is there something special at the end? (We know that index validation is a distinct and potentially expensive step, but I don’t think that’s what we’re talking about here.)

- Are there general performance opportunities for index backfills? Hypotheses:
  - Is there some quadratic behavior?
  - Are we operating on too-small pieces?
  - Can we parallelize intra-node?
  - Can we distribute better?

Further goals are better observability, and more accurate progress calculation.

cc @pbardea @ajwerner (feel free to add or edit)

gz#9008

gz#9028

Epic CRDB-8816

gz#8886

gz#8907

Jira issue: CRDB-8594

**Comment by ajwerner** [comment]
I don't think we observed anything quadratic here. I think it's that we aren't able to build large SSTs when tables are large and we have no parallelism. Imagine that we have 50k ranges and we have a buffer we fill of 512 MiB. That's going to be 10 KiB per range on average. We'll send these as SSTs with zero parallelism. We observed anecdotally that we were able to push about ~100-200 of these per node per second which amounts to ~1MiB/s. In practice it seemed even worse due to straggler nodes leading to higher execution latency. Perhaps there are other reasons and knobs why we were building even smaller SSTs. Note that none of these are big enough to be ingested as SSTs.

What I feel we need is parallelism sending requests when the sizes get small.

**Comment by nvanbenschoten** [comment]
> I don't think we observed anything quadratic here.

Don't we also expect the calls to `checkForKeyCollisions` during `AddSSTableRequests` to be quadratic? This is what the idea in https://github.com/cockroachdb/cockroach/issues/66410 is hoping to combat.

**Comment by ajwerner** [comment]
> > I don't think we observed anything quadratic here.
> 
> Don't we also expect the calls to `checkForKeyCollisions` during `AddSSTableRequests` to be quadratic? This is what the idea in #66410 is hoping to combat.

This is news to me but I don't think it features actively here. Namely we know that the SSTs in question here are tiny (100% under 400KiB) and we have strong evidence that the latency is less than 10ms. The evidence for the latency being ~10ms is that we saw that each sender was single-threaded and we saw them being sent at around 100/s per sender. 

I guess what I'm saying is that maybe what you're saying is a problem and a source of the latency we do observe, but even if we could send these requests two or three times faster, it still would not be fast.

My sense is that if we could keep a pool of sending goroutines running which have a certain byte budget for the number of bytes we have in flight, I think we'd do a lot better at the end of these large backfills.

**Comment by ajwerner** [comment]
![image](https://user-images.githubusercontent.com/1839234/125530260-480b5007-1826-4bf9-9dc7-416f04573248.png)

Maybe if we didn't have this quadratic behavior then, best case, we could push ~1000/s. Still, a pool of 16 goroutines which are allowed at most 1MiB outstanding or something would probably be an even bigger improvement.

**Comment by nvanbenschoten** [comment]
> Still, a pool of 16 goroutines which are allowed at most 1MiB outstanding or something would probably be an even bigger improvement.

Dynamic sender concurrency with a fixed data rate makes a lot of sense to me.

For completeness, we confirmed that we did not see a correlated increase in read amplification during this slowdown, right?

**Comment by ajwerner** [comment]
We did, in fact, see a marked increase in read amplification on two nodes. How do you figure that ties in?

**Comment by ajwerner** [comment]
Is the thinking that the scale of this query hides periods of actually very slow requests on the nodes with the high read amplification? I buy that. Even so, it still seems like parallelism is worthwhile.

**Comment by ajwerner** [comment]
![image (1)](https://user-images.githubusercontent.com/1839234/125532804-07eff760-5cc9-4af9-ba79-0a680f5f5f7d.png)

**Comment by nvanbenschoten** [comment]
Well read amplification is going to make this `checkForKeyCollisions` operation that's run on every `AddSSTable` request even less efficient, slowing it down proportionally to the read amplification factor. In other cases where I've seen index backfills and IMPORTs slow to a crawl, one of the nodes saw extreme read amplification growth (potentially due to https://github.com/cockroachdb/pebble/issues/1181) and its calls to `checkForKeyCollisions` began dominating CPU profiles.

I don't think that's quite what we're seeing here though, as read amplification spikes but gets back under control around 4:00, _before_ we hit the long tail of `AddSSTable` requests.

**Comment by ajwerner** [comment]
IIRC both the disk IO and CPU on these machines was far from being fully utilized, including the machines with high read amp.

**Comment by pbardea** [comment]
I also wanted to add that backfills [don't set](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/rowexec/indexbackfiller.go#L191-L198) `DisallowShadowing`, so `checkForKeyCollisions` shouldn't be called at all during the backfill.

**Comment by miretskiy** [comment]
Just spitballing: during index backfill we produce a lot of sstables; we observed with one of the clients that while disk IO and CPU were okay on the machine, they still saw significant read amplification and that was due to pebble encryption at rest serialization during manifest updates (to add sstable, read potentially large manifest, decrypt it, decode proto, add sst, encode proto, encrypt, write manifest); see https://github.com/cockroachlabs/support/issues/1098

**Comment by dt** [comment]
This is an old issue with no updates in years. Backfills have been rewritten extensively (e.g. offline pre-backfil now) since it was filed, so I'm closing this one and if new problems are observed in the new implementations, we can file a new issue for them
